static/js/publisher/market/market.js

```diff
@@ -35,9 +47,110 @@ function initFormNotification(formElId, notificationElId) {
   }
 }
 
+const allowedKeys = ['title', 'summary', 'description', 'images', 'website', 'contact'];
+
+// TODO: test
+function diffFormData(initialState, state) {
+  const diff = {};
+
+  for (let key of allowedKeys) {
+    // images is an array of objects so compare stringified version
+    if (key === 'images') {
+      const images = state[key]
+        // remove images to delete from the diff
+        .filter(image => image.status !== 'delete')
+        // ignore selected status when comparing
+        .map(image => {
+          delete image.selected;
+          return image;
+        });
+
+      if (JSON.stringify(initialState[key]) !== JSON.stringify(images)) {
+        diff[key] = images;
+      }
+    } else {
+      if (initialState[key] !== state[key]) {
+        diff[key] = state[key];
+      }
+    }
+  }
+
+  // only return diff when there are any changes
+  return Object.keys(diff).length > 0 ? diff : null;
+}
+
+// TODO: test
+function updateState(state, values) {
+  if (values) {
+    // if values can be iterated on (like FormData)
+    if (values.forEach) {
+      values.forEach((value, key) => {
+        if (allowedKeys.includes(key)) {
+          state[key] = value;
+        }
+      });
+    // else if it's just a plain object
+    } else {
+      for (let key in values) {
+        if (allowedKeys.includes(key)) {
+          state[key] = values[key];
+        }
+      }
+    }
+  }
+}
+
+function initForm(config, initialState) {
+  const marketForm = document.getElementById(config.form);
+  let state = JSON.parse(JSON.stringify(initialState));
+
+  const stateInput = document.createElement('input');
+  stateInput.type = "hidden";
+  stateInput.name = "state";
+
+  marketForm.appendChild(stateInput);
+
+  const diffInput = document.createElement('input');
+  diffInput.type = "hidden";
+  diffInput.name = "changes";
+
+  marketForm.appendChild(diffInput);
+
+  initSnapIconEdit(config.snapIconImage, config.snapIconInput, state);
+  initFormNotification(config.form, config.formNotification);
+  initSnapScreenshotsEdit(
+    config.screenshotsToolbar,
+    config.screenshotsWrapper,
+    state
+  );
+
+  // when anything is changed update the state
+  marketForm.addEventListener('change', function() {
+    updateState(state, new FormData(marketForm));
+  });
+
+  document.querySelector('.js-market-submit').addEventListener('click', function(event) {
```

Comment:
Is this not a submit event on the form for a reason? What happens when people press return?

Reply:
Currently it reverts the form!?

Reply:
Yes, both 'Revert' and 'Apply' are submitting, that's why 'Enter' picks up first and 'Reverts' it.

I wanted to have it on `submit`, but because we have 2 submit buttons I wasn't able to tell the difference in `submit` event which button was clicked (without hacks), so `click` seemed to be more reasonable.

Would be good to make 'Enter' do a submit. I guess it would require changing button order maybe, so that 'Apply' is first in markup.

Reply:
Does making the revert a normal `button` and apply a `type="submit"` button fix that? It would be more semantic too.